### DISCUSSION TITLE
Can't mock module imported inside dependency

### DISCUSSION THREAD

**Discussion by joao8lopes** [vitest-dev/vitest#6012] [Q&A]
Hi team 👋🏽 

I'm encountering an issue where a module imported within a dependency is not mocked as expected during testing.

Example:

**handler.ts**

```
import { apiHandler } from 'api-handler';

const hello = () => {
  return apiHandler();
};
```

**api-handler.ts**
```
import { Log } from 'log'

const apiHandler = () => {
  const log = new Log();
  log.info('This is my log');
}
```

in the unit test, **index.test.ts**
```
vi.mock('log');

test('my test', async () => {
  await hello();
  // assertions 
})
```

When running `npx vitest`, I still see the log `This is my log in the console`. I expected this not to happen since 'log' is being mocked. The same issue occurs when trying to mock the implementation of a function that's being created and used inside `api-handler` where the mocked function is not called and instead, the original function is.

Is there a reason why the mock isn't being applied as expected?

Thank you!

**Comment by joao8lopes** [comment]
bump

**Comment by StaceyD22** [comment]
Unless you provide a factory function, vitest will use magic ([algorithms](https://vitest.dev/guide/mocking.html#automocking-algorithm)) to automock, which might not be what you desire. If you don't want to see that log, pass in a factory function.

**Reply by stantonwjones** [reply]
if a factory function is not provided, it should mock all named exports according to the documentation. This is definitely a bug and I'll probably have to shoe in a third party library to handle this until it's fixed

**Reply by StaceyD22** [reply]
It is not a bug. Did you follow that link. No factory function was passed, so vitest deeply cloned the class…. The simple solution is to pass a factory function! 

> Automocking Algorithm
If your code is importing a mocked module, without any associated __mocks__ file or factory for this module, Vitest will mock the module itself by invoking it and mocking every export.

The following principles apply

All arrays will be emptied
All primitives and collections will stay the same
All objects will be deeply cloned
All instances of classes and their prototypes will be deeply cloned

**Reply by stantonwjones** [reply]
I guess I assumed _**'by invoking it and mocking every export'**_ implied that the exported functions would be mocked.

However, for due diligence I tried passing a factory function. I've already passed a generic object defining the mocked methods I wanted returned instead of the original import. I already tried specifying the __mocks__ folder/file alongside the original module.

Each of those failed in the manner described by lopes

**Reply by stantonwjones** [reply]
The issue I'm encountering may be due to some dynamic imports bypassing with whatever vitest is doing to override the import statement. For now, I'm using the environment variables in the vite config to manually change behavior in the test suite.
```
// MyModuleIWantToMock.js
export function cantBeMocked() {
  if (process.env.runtime_env == "test") return {test: 'data'}
  else {
    // Do normal behavior
  }
}
```
 I'll continue to investigate the root cause
